File: PSet6/sentimental-credit/credit.py
from cs50 import get_int
from math import ceil, log10


def main():

    # Require a number and convert to string
    number = str(get_int("Number: "))
    length = len(number)
    valid = check_credit_card_number(number)
    start = int(number[0:2])

    if valid:

        # Check the first two digits only when the number is valid
        if length == 15 and (start == 34 or start == 37):
            print("AMEX")
        elif length == 16 and (start >= 51 and start <= 55):
            print("MASTERCARD")
        elif (length == 13 or length == 16) and (start >= 40 and start <= 49):
            print("VISA")
        else:
            print("INVALID")

    else:
        print("INVALID")


# Returns if a number is valid
def check_credit_card_number(number):

    sum = 0

    # Add digits which should be multiplied by 2
    for digit in number[-2:0:2]:
        sum += int(digit) * 2

    # Add other digits
    for digit in number[-2:0:2]:
        sum += int(digit)

    return sum % 10 == 0


if __name__ == "__main__":
    main()

File: PSet1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool check_credit_card_number(long number);

int main(void)
{
    long number = get_long("Number: ");
    bool valid = check_credit_card_number(number);
    int length = ceil(log10(number)); // Number of digits
    int start = number / pow(10, length - 2);

    if (valid)
    {
        // Check the first two digits only when the number is valid
        if (length == 15 && (start == 34 || start == 37))
        {
            printf("AMEX\n");
        }
        else if (length == 16 && (start >= 51 && start <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((length == 13 || length == 16) && (start >= 40 && start <= 49))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Returns if the given number is a valid credit card number
bool check_credit_card_number(long number)
{
    if (number <= 0)
    {
        return false;
    }

    bool mult_by2 = false; // indicates if the current digit needs to be multiplied by 2
    int sum = 0;

    while (number != 0)
    {
        int digit = number % 10;
        number = floor(number / 10);

        // Each digit is obtained (starting from the last)
        if (mult_by2)
        {
            digit *= 2;

            // The maximum value of a digit multiplied by 2 is 18 = 9 * 2
            // When the result has two digits, add 1 to the sum, and 10 is subtracted from the result
            if (digit >= 10)
            {
                sum += 1;
                digit -= 10;
            }
            sum += digit;
        }
        else
        {
            sum += digit;
        }
        mult_by2 = !mult_by2;
    }

    return sum % 10 == 0;
}